Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+
+An implementing program must not process any statements inside a block that is not the active block.
+An implementing program must only have, at most, one active block.
+If no blocks expression evaluates to true then no block must be processed.
+
+#### Ambiguity
+
+Because the flow control commands allow for potentially multiple commands and their blocks to be the true one, the implementing program must select them in a top down approach when there is conflicting flow.
+For example take the following flow control:
+
+```yarn
+<<if false>>
+    if-scope
+<<elseif true>>
+    elseif-1-scope
+<<elseif true>>
+    elseif-2-scope
+<<else>>
+    else-scope
+<<endif>>
+```
+
+Both of the elseif commands expressions evaluate to true, so either ones attached block could be run and seen to be correct.
+However because one is above the other the block with `elseif-1-scope` dialogue inside would be the selected one.
+The implementing program should attempt to identify these scenarios however and alert the writer.
+
+### Options
+
+_Options_ are the means by which Yarn can present dialogue choices to the game and much as with flow control are an element that spans multiple lines.
+Options are comprised of one or more option lines.
+An _option line_ represents a single choice in an option, and are comprised of three parts: the keyword, the dialogue, the conditional in that order.
+
+The _keyword_ is how the implementing program can tell a line is part of an option instead of dialogue and is the symbols `->`.
+There must be at least one whitespace between the keyword and the next element, the dialogue.
+The _dialogue_ is a normal line of dialogue following all rules associated with that.
+
+The _conditional_ is a command which controls the specifics of the validity of the option.
+The conditional allows for additional data about the option to be bundled along with the line.
+The intent of the conditional is for the game to know which lines are currently valid for selection and which are not without forcing the writer to duplicate options inside of a control flow.
+The conditional's syntax is identical to the if command and follows all rules there, but as it is not part of flow control must not have an accompanying endif or attached block.
+The conditional must be an optional component of the option line.
+
+As the intention of options is to provide choice when options are encountered the implementing program must halt further progress through the node until an option has been selected.
+Each option must be provided in the order they are written in the node.
+The mechanism by which an option line is chosen is unspecified.
+Only a single option line must be chosen.
+
+As the conditional controls the validity of the option for selection this information must be provided by implementing program to the parts of the game that makes the selections.
+The implementing program must not restrict the selection of invalid options.
```

Comment:
Pardon? Is "valid options" meant here?

Reply:
I think it's saying that when given a list of optionals with defined conditionals, The Yarn implementation should be given the full list of conditionals, even invalid ones, and then the 'game' side should be responsible for handling invalid options (see line below)

...which seems very strange to me, given the if/elseif etc blocks above basically prevent the whole chunk inside the block from being processed at all.

It would make a lot more sense to me if the option syntax was ultimately functionally equivalent to the if/elseif described above, i.e. if the conditional evaluates false, the option is essentially not 'processed' and therefore not even provided to the rest of the game.

Reply:
That's how Chatterbox does it at the moment, but occasionally RPGs will show you options that you can't take because you lack the required stats. I'd imagine that this part of the Yarn spec is keeping this functionality as a possibility

Reply:
Hmm, I see what you mean, though it's weird that the only way to accomplish that is via this shortcut option syntax, and not regular options. And how are the other components of the program supposed to parse the yarn conditional to check validity? I imagine the spec should say something about how the yarn implementation should evaluate and communicate the option validity to the other components...

Reply:
Yeah, what's currently proposed (and is being added to Yarn Spinner 2.0, though it's not yet done) is a change to the semantics of `<<conditions>>` at the end of options; currently, the condition controls whether the option is sent to the game at all, but in a future beta, all options will be sent, along with information on whether the condition passed or not. This is to support use cases like "this option would have been available if you'd had enough skill points". 

See https://github.com/YarnSpinnerTool/YarnSpinner/issues/265 for full discussion on this.

Reply:
> Hmm, I see what you mean, though it's weird that the only way to accomplish that is via this shortcut option syntax, and not regular options.

I think there is some confusion around this and I want to rewrite it to say something like "options must be sent over with the line of text and a boolean indicating their validity" or something to that effect.
But that only shortcuts can do this is because the original syntax for options is going away (see #264 for the discussion around this).